test/test_broker.hpp

```diff
@@ -1052,22 +1106,19 @@ class test_broker {
 
                     e.con.reset();
 
-                    auto const& sei_opt = e.session_expiry_interval;
-                    if (sei_opt && sei_opt.value() !=
-                        std::chrono::seconds(MQTT_NS::session_never_expire)) {
-                        e.tim_session_expiry = std::make_shared<as::steady_timer>(ioc_);
-                        e.tim_session_expiry->expires_after(sei_opt.value());
+                    if (e.session_expiry_interval && e.session_expiry_interval.value() != std::chrono::seconds(MQTT_NS::session_never_expire)) {
+                        e.tim_session_expiry = std::make_shared<as::steady_timer>(ioc_, e.session_expiry_interval.value());
                         e.tim_session_expiry->async_wait(
-                            [&, wp = std::weak_ptr<as::steady_timer>(e.tim_session_expiry)]
-                            (MQTT_NS::error_code ec) {
-                                auto sp = wp.lock();
-                                if (!ec) {
-                                    auto& idx = sessions_.get<tag_tim>();
-                                    idx.erase(sp);
+                            [&, wp = std::weak_ptr<as::steady_timer>(e.tim_session_expiry)](MQTT_NS::error_code ec) {
```

Comment:
```suggestion
                            [this, wp = std::weak_ptr<as::steady_timer>(e.tim_session_expiry)](MQTT_NS::error_code ec) {
```

It is the original code problem. Implicit captiore `&` shouldn't be used with async function's callback.
See https://github.com/redboltz/mqtt_cpp/wiki/Coding-Rules#details-of-lambda-expression